provisionerd/runner.go

```diff
@@ -0,0 +1,872 @@
+package provisionerd
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"path/filepath"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/spf13/afero"
+	"golang.org/x/xerrors"
+
+	"cdr.dev/slog"
+
+	"github.com/coder/coder/provisionerd/proto"
+	sdkproto "github.com/coder/coder/provisionersdk/proto"
+)
+
+type jobRunner interface {
+	start()
+	cancel()
+	isDone() <-chan any
+	fail(ctx context.Context, f *proto.FailedJob) error
+	forceStop()
+}
+
+type runner struct {
+	job                 *proto.AcquiredJob
+	sender              messageSender
+	logger              slog.Logger
+	filesystem          afero.Fs
+	workDirectory       string
+	provisioner         sdkproto.DRPCProvisionerClient
+	updateInterval      time.Duration
+	forceCancelInterval time.Duration
+
+	// mutex controls access to all the following variables.
+	mutex *sync.Mutex
+	// used to wait for the failedJob or completedJob to be populated
+	cond *sync.Cond
+	// closed when the job ends.
+	done         chan any
+	failedJob    *proto.FailedJob
+	completedJob *proto.CompletedJob
+	// active as long as we are not canceled
+	gracefulContext context.Context
+	cancelFunc      context.CancelFunc
+	// active as long as we haven't been force stopped
+	forceStopContext context.Context
+	forceStopFunc    context.CancelFunc
+	// setting this false signals that no more messages about this job should be sent.  Usually this means that a
+	// terminal message like FailedJob or CompletedJob has been sent, but if we are force canceled, we may set this
+	// false and not send one.
```

Comment:
I should reword a little bit.

If you call `cancel()` we attempt to gracefully cancel, and send a terminal message when that's done.

If you call `fail()` we attempt to send FailedJob, but only on the provided context, so we might give up before the message is successfully sent.

If you call `forceStop()` we immediately stop and don't try to send any messages.

In the latter two cases, you might end up in a situation where we are done processing but haven't sent any message.  This is for e.g. stopping provisionerd when it can't reach coderd.  Eventually we give up even if we haven't sent the terminal message.